README.md
# alpaka

[카프카](https://kafka.apache.org/)는 분산형 데이터 스트리밍 플랫폼으로 비교적 설정이 복잡하고 함께 사용하는 서비스도 다양해 설치가 좀 까다롭다. 알파카는 Kubernetes 클러스터에 Kafka 및 관련 패키지를 쉽게 배포하기 위한 Helm 차트이다 (차트는 설치 패키지와 비슷한 개념으로 여기서는 혼용하겠다). 다음과 같은 외부 패키지를 포함한다:

- Kafka
- Zookeeper
- Kafka 용 JDBC 소스 커넥터 + S3 싱크 커넥터
- ksqlDB
- UI for Kafka
- Prometheus (+ KMinion)
- Grafana (+ 각종 대쉬보드)
- Grafana Loki 
- Kubernetes Dashboard

추가적으로 테스트 및 운영에 필요한 툴 컨테이너가 설치된다.

## 사전 준비

OS 별로 도커를 사용할 수 있도록 준비하고, 여기에 추가적으로 [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/) 과 [Helm](https://helm.sh/docs/intro/install/) 가 설치되어 있어야 한다. 각각의 링크를 참고하여 진행하자.

다음으로 사용할 쿠버네티스 배포판을 선택해야 한다. 개발용으로는 로컬 쿠버네티스 배포판을, 프로덕션 용도로는 클라우드 또는 IDC 용 배포판을 이용할 수 있다. 알파카는 로컬용으로 [minikube](https://minikube.sigs.k8s.io/docs/), [k3s](https://k3s.io/) 및 [k3d](https://k3d.io/v5.4.6/) 를, 프로덕션용으로 [AWS EKS](https://aws.amazon.com/ko/eks/) 을 위한 설정 파일을 기본으로 제공한다. 

다른 배포판 환경에도 조금만 응용하면 무리없이 적용할 수 있을 것이다.

### 로컬 쿠버네티스 배포판 관련

로컬 쿠버네티스 환경에서는 메모리가 너무 작으면 파드가 죽을 수 있고, 디스크 용량이 너무 작으면 PVC 할당이 안될 수 있으니 주의하자. 

#### minikube 이용시

코어 4개, 메모리 8GB, 디스크 40GB 예:
```
minikube start --cpus=4 --memory=10g --disk-size=40g
```

#### k3d 이용시 

워커노드 2 대, 노드별 메모리 8GB, Ingress 이용의 예:

```
K3D_FIX_DNS=1 k3d cluster create -p "80:80@loadbalancer" --agents 2 --agents-memory=8gb
```

> K3D_FIX_DNS 가 없으면 생성된 클러스터내 파드에서 인터넷 접속 안되는 문제가 있다.
> ( https://github.com/k3d-io/k3d/issues/209 )


### 클라우드 쿠버네티스 환경 (AWS EKS) 관련

여기서는 AWS EKS (관리형 쿠버네티스 서비스) 기준으로 설명한다.

#### 클러스터 만들기 

아래는 클러스터 이름 `prod`, EC2 `m5.xlarge` 타입 4대, 디스크 100GB 최대 4대인 예이다. 참고하여 필요한 사양의 클러스터를 만들도록 하자.

```bash
eksctl create cluster \
--name prod \
--nodegroup-name xlarge \
--node-type m5.xlarge \
--node-volume-size 100 \
--nodes 1 \
--nodes-min 1 \
--nodes-max 4
```

> AWS EKS 클러스터의 경우 사용을 마쳤으면 꼭 제거해 비용을 절감하도록 하자.
> 
> ```
> eksctl delete cluster --name prod
> ```

#### 스토리지 프로비저닝 준비 

클러스터 생성후 PVC 를 통한 스토리지 할당을 받기 위해 아래 작업이 필요하다.
( 참고 : [Creating an IAM OIDC provider for your cluster](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) )

1. 기존 OIDC 존재 확인
```bash
oidc_id=$(aws eks describe-cluster --name prod --query "cluster.identity.oidc.issuer" --output text | cut -d '/' -f 5)
```

2. 클러스터 ID 의 IAM OIDC 프로바이더가 계정에 있는지 확인 

```bash
aws iam list-open-id-connect-providers | grep $oidc_id
```

3. 결과가 없으면 IAM OIDC 아이덴터티 생성
```bash
eksctl utils associate-iam-oidc-provider --cluster prod --approve
```

4. 서비스 계정을 위한 EBS CSI 드라이버용 IAM Role 생성
( 참고 : [Creating the Amazon EBS CSI driver IAM role for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/csi-iam-role.html) )

```bash
eksctl create iamserviceaccount \
  --name ebs-csi-controller-sa \
  --namespace kube-system \
  --cluster prod \
  --attach-policy-arn arn:aws:iam::aws:policy/service-role/AmazonEBSCSIDriverPolicy \
  --approve \
  --role-only \
  --role-name AmazonEKS_EBS_CSI_DriverRole
```

5. EKS 애드온으로 EBS CSI 드라이버 설치하기
( 참고 : [Managing the Amazon EBS CSI driver as an Amazon EKS add-on](https://docs.aws.amazon.com/eks/latest/userguide/managing-ebs-csi.html) )
```bash
eksctl create addon --name aws-ebs-csi-driver --cluster prod --service-account-role-arn arn:aws:iam::<AWS 계정 번호>:role/AmazonEKS_EBS_CSI_DriverRole --force
```

#### EKS 용 인그레스 (Ingress) 준비 

알파카로 설치되는 다양한 서비스의 웹페이지를 외부에서 접근하기 위해서 인그레스가 필요하다. EKS 를 이용하는 경우 인그레스를 사용하기 전에 클러스터 생성후 아래 작업이 필요하다.

AWS 로드밸런서 컨트롤러 설치
( 참고 : [Installing the AWS Load Balancer Controller add-on](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) )

1. IAM Policy 생성 ( 정책 이름 : `AWSLoadBalancerControllerIAMPolicy` )

`etc/eks/` 디렉토리로 이동 후 다음과 같이 실행한다.
```bash
aws iam create-policy \
    --policy-name AWSLoadBalancerControllerIAMPolicy \
    --policy-document file://iam_policy.json
```

2. IAM Role 만들기( 역할 이름 : `AmazonEKSLoadBalancerControllerRole` )
```bash
eksctl create iamserviceaccount \
  --cluster=prod \
  --namespace=kube-system \
  --name=aws-load-balancer-controller \
  --role-name "AmazonEKSLoadBalancerControllerRole" \
  --attach-policy-arn=arn:aws:iam::<AWS 계정 번호>:policy/AWSLoadBalancerControllerIAMPolicy \
  --approve
```

4. Helm 으로 EKS 용 Load Balancer Controller 설치

차트 저장소 추가
```bash
helm repo add eks https://aws.github.io/eks-charts
```

로컬 저장소 갱신
```bash
helm repo update
```

차트 설치
```bash
helm install aws-load-balancer-controller eks/aws-load-balancer-controller \
  -n kube-system \
  --set clusterName=prod \
  --set serviceAccount.create=false \
  --set serviceAccount.name=aws-load-balancer-controller 
```

> 참고: 배포된 차트의 보안 업데이트
>
> ```kubectl apply -k "github.com/aws/eks-charts/stable/aws-load-balancer-controller/crds?ref=master"```

설치 확인

```kubectl get deployment -n kube-system aws-load-balancer-controller```

이제 AWS EKS 에서 인그레스를 이용할 준비가 되었다. AWS EKS 의 경우 도메인 명으로 접속하려면 퍼블릭 도메인과 ACM 인증서가 필요하기에, 여기서는 도메인 없이 포트로 서비스를 구분하여 사용하는 것으로 설명하겠다. `configs/_eks.yaml` 설정 파일을 보면 이를 위해 설정 파일에서 서비스 별로 인그레스를 서로 다른 포트로 요청하는 것을 확인할 수 있다. 

> 퍼블릭 도메인을 이용하는 경우:
> AWS ACM 으로 가서 퍼블릭 도메인을 위한 인증서를 만들어 주고 그 ARN 을 아래와 같이 `annotations` 아래에 기재하여야 한다.
> ```
> alb.ingress.kubernetes.io/certificate-arn: arn:aws:acm:ap-northeast-2:111122223333:certificate/11fceb32-9cc2-4b45-934f-c8903e4f9e12 
> alb.ingress.kubernetes.io/ssl-redirect: '443'
> ```

### 부속 컨테이너 이미지  

외부 패키지를 이용할 때는 그쪽에서 공개한 공식 컨테이너 이미지를 이용하면 된다. 추가적으로 알파카가 사용하는 부속 컨테이너 이미지는 직접 만들어 주어야 할 경우가 있기에 여기서 소개한다. 현재 이미지는 툴 및 DB 커넥터용의 두 가지가 있다.

#### 툴 컨테이너 이미지

서비스를 위한 컨테이너 이미지는 용량, 속도, 보안 등의 이슈로 대부분 서비스에 꼭 필요한 실행 파일만 설치되어 있다. 그런데 패키지를 설치하고 운용하다 보면 여러가지 문제가 발생하기에, 이를 찾고 고치기 위한 실행파일이 설치된 이미지가 있으면 편리하다. 알파카에서는 이것을 툴 컨테이너 이미지라고 부른다. 

여기에는 vim, kcat, jq, 파이썬 및 DNS 유틸 등이 설치되어 있기에, 클러스터 내 다른 컨테이너의 동작을 확인하거나 네트워크 이슈를 찾는 등 다양한 작업을 할 수 있다. 기본 툴 이미지는 도커 허브의 [haje01/alpaka-tool](https://hub.docker.com/repository/docker/haje01/alpaka-tool/general) 로 올라가 있기에 이것을 사용하면 된다. 


> 만약 독자적인 툴 이미지가 필요하면 `images/Dockerfile.tool` 및 `images/build_tool.sh` 파일을 참고하여 만들도록 하자.

#### 커넥트 이미지 

카프카의 커넥터 (Connector) 는 다양한 데이터 소스에서 데이터를 가져오거나, 외부로 내보내는 데 사용되는 일종의 플러그인이다. 커넥터는 카프카 커넥트 (Connect) 파드에 등록되어 동작하는데, 이를 위해 알파카에서는 기본적으로 두 가지 이미지를 이용한다. 하나는 외부 데이터를 카프카로 가져오는 소스 커넥터를 위한 이미지로 [Confluent JDBC 커넥터](https://docs.confluent.io/kafka-connectors/jdbc/current/index.html) 및 [Debezium](https://debezium.io/) 등이 설치되어 있다. 다른 하나는 카프카 토픽의 내용을 외부로 내보내기 위한 싱크 커넥터를 위한 이미지로 현재는 S3 Sink 가 설치되어 있다.


소스 커넥트용 이미지는 도커 허브의 [haje01/kafka-srccon](https://hub.docker.com/repository/docker/haje01/kafka-srccon/general) 에, 싱크 커넥트용 이미지는 [haje01/kafka-srccon](https://hub.docker.com/repository/docker/haje01/kafka-srccon/general) 에 있는 것을 사용하면 된다. 

> 만약 기본 이미지외 독자적인 커넥트 이미지가 필요하면 `images/Dockerfile.srccon` 및 `images/build_srccon.sh` 파일을 참고하여 만들고 설정을 바꾸어서 이용하면 된다.


## 설정 파일 만들기 

설치를 위해서는 먼저 설정 파일이 필요하다. `configs/` 디렉토리에 아래와 같은 샘플 설정 파일이 있으니 참고하여 자신의 필요에 맞는 설정 파일을 만들어 사용한다 (샘플 설정 파일은 구분하기 쉽게 `_` 로 시작한다):
- `_mkb.yaml` - minikube 용
- `_k3s.yaml` - k3s 용
- `_k3d.yaml` - k3d 용
- `_eks.yaml_` - eks 용

> 여기서는 참고용으로 쿠버네티스 배포판별 설정 파일을 만들었지만 꼭 이렇게 할 필요는 없다. 실제로는 한 번 선택한 쿠버네티스 배포판 자주 바뀌지 않기에 개발/테스트/라이브 등의 용도별로 설정 파일을 만드는 것이 더 적합할 것이다.

> `alpaka/values.yaml` 는 차트에서 사용하는 기본 변수값을 담고 있다. 위의 파일들과 함께 참고하여 커스텀 설정 파일을 만들 수 있겠다.

설정 파일은 [YAML](https://yaml.org/) 형식의 파일로, 최상위 블럭만 표시하면 다음과 같은 구조를 가진다.

```yaml
k8s_dist:       # 쿠버네티스 배포판 종류 (minikube, k3s, k3d, eks 중 하나)

kafka:          # 카프카 설정 

kafka_connect:  # 카프카 커넥트 및 커넥터 설정 

ui4kafka:       # UI for Kafka 설정

k8dashboard:    # 쿠버네티스 대쉬보드 설정 

prometheus:     # 프로메테우스 설정

grafana:        # 그라파나 설정 

loki:           # 그라파나 로키 설정

kminion:        # 그라파나 용 KMinion 대쉬보드 설정 

ingress:        # 공용 인그레스 설정

init:           # 클러스터 설치 후 초기화 설정 

test:           # 클러스터 설치 후 테스트 설정 
```

`k8s_dist`, `kminion`, `ingress`, `kafka_connect`, `init` 그리고 `test` 는 알파카 자체적인 설정 블럭이다. `kafka`, `ui4kafka`, `k8dashboard`, `prometheus`, `grafana` 는 알파카가 의존하는 외부 패키지를 위한 설정 블럭이다.

지금부터는 테스트를 위한 간단한 설정 파일을 만들어 가면서 설정 파일 작성 방법을 소개하겠다. 이 설정 파일은 로컬에서 minikube 배포판을 이용하고, 설정 파일은 `configs/_mkb.yaml` 샘플을 참고하여 `configs/mymkb.yaml` 에 저장하는 것으로 가정하겠다. 파일의 최초 내용은 아래와 같다.

```yaml
k8s_dist: minikube
```

이 설정 파일이 완성되면 다음과 같이 설치하게 될 것이다.

```
helm install -f configs/mymkb.yaml my alpaka/
```

이 경우 배포 이름은 `my` 이 된다. 

이제 각 설정 블럭을 차례대로 설명하겠다. 알파카 차트나 알파카 변수 파일 `alpaka/values.yaml` 에 기본값이 지정되어 있고 설정 파일에 관련 내용이 없으면 기본값이 이용된다.

### 카프카 설정 

카프카 클러스터 및 주키퍼 관련 내용을 여기서 설정한다. 대략의 구조는 아래와 같다.

```yaml
kafka: 
  replicaCount: 1              # 카프카 브로커의 수. HA 구성을 위해서는 3 을 추천 
  defaultReplicationFactor: 1  # 토픽 복제 수. 브로커 수가 3 인 경우 2 를 추천 
  numPartitions: 8             # 토픽별 파티션 수
```

기본 값은 브로커 1 대, 파티션 수 8 개로 이대로 사용하고자 하면 별도의 `kafka` 블럭을 지정하지 않아도 카프카 클러스터가 만들어진다.

카프카는 [bitnami 의 kafka Helm 차트](https://bitnami.com/stack/kafka/helm) 를 이용하였다. 로컬 Helm 저장소에 등록하였다면 차트의 기본값은 다음처럼 확인할 수 있다.

```
helm show values bitnami/kafka
```

### 카프카 커넥트 설정 

카프카 커넥터는 데이터를 카프카로 가져오거나 외부로 내보내기 위한 일종의 플러그인이다. 커넥트는 커넥터를 기동하기 위한 장비로 생각하면 간단하다.

커넥트 설정은 꽤 복잡한데 대략적인 구조를 표시하면 아래와 같다.

```yaml
kafka_connect:
  # 커넥트 리스트
  connects:
    # 개별 커넥트 정보 
    - type:    # 커넥트 타입
      values:  # 커넥트 내에서 공유되는 변수값
    # 커넥터 그룹 리스트 
    connector_groups:
      # 개별 커넥터 그룹 정보 
      - name:    # 커넥터 그룹 이름 
        values:  # 커넥터 그룹 내에서 공유되는 변수값
        common:  # 커넥터 그룹 내 공통 설정
        # 개별 커넥터 리스트 
        connectors:  
          - name:    # 커넥터 이름
            config:  # 커넥터 개별 설정 
```

최초로 `connect` 블록 아래에 하나 이상의 커넥트 정보를 기술할 수 있다. 커넥트는 같은 커넥터 라이브러리를 사용하는 서비스를 하나 이상 기동하기 위한 장비의 구분 정도로 생각하자. 

각 커넥트에는 `values` 블럭이 나올 수 있는데, 커넥터 내에서 공유되는 변수를 기술하는데 사용된다.

커넥터는 크게 비슷한 설정을 가지는 것들끼리 그룹지을 수 있는데, 이것을 위해 `connector_groups` 블럭이 존재하며 그 아래 커넥터 그룹 리스트가 나온다.

각 커넥터 그룹에는 `values` 와 `common` 블럭이 있는데, 각각 커넥터 그룹 내에서 공유되는 변수 및 공통 설정을 기술하는 데 사용된다.

아래는 DB 에서 데이터를 가져오기 위한 JDCB 소스 커넥트를 위한 구체적인 예로, 여기서는 MySQL 에 있는 `mydb` 라는 DB 에서 로그성 테이블 `log1` 과 `log2` 그리고 코드성 테이블 `code` 를 가져오는 경우를 가정하였다.

> 이 예는 어디까지나 설명을 위한 것으로 실제 커넥터 설정은 좀 더 상세한 설정이 필요하다.

```yaml
kafka_connect:
  enabled: true       # 커넥트를 사용하면 true.
  # 커넥트 정보
  connects:           # 이 블록아래 커넥트를 종류별로 기술 
  - type: srccon        # 커넥트 타입 정보. 여기서는 JDBC 소스 커넥터 타입 
    replicaCount: 1      # 커넥트 파드 수
    # 컨테이너 정보 
    container:                     
      image: "haje01/kafka-srccon"   # 컨테이너 이미지 
      tag: 0.0.5                     # 이미지 태그 
      pullPolicy: IfNotPresent       # 이지미 풀 정책 
    timezone: Asia/Seoul # 커넥트 파드의 타임존 
    # 필요한 경우 컨테이너에 건낼 환경변수 지정
    envs: []
    # 커넥트 레벨에서 공유될 변수 값 리스트 
    values:                    
      db_host: mysql-db-addr      # DBMS 호스트 주소
      db_port: 3306               # DBMS 접속 포트
      db_name: mydb               # DB 이름 
      db_user: myuser             # DB 유저
      db_pass: mypass             # DB 유저 암호 
    # 커넥터 그룹 리스트 
    connector_groups:          
    # DB 내 로그성 테이블을 가져오기 위한 커넥터 그룹 
    - name: mydb_log  # 커넥터 그룹 이름 
      # 그룹내 커넥터 공통 설정
      common:
        connector.class: io.confluent.connect.jdbc.JdbcSourceConnector    # 커넥터 클래스 
        tasks.max: "1"   # 커넥터 동시성 수 
        connection.url: "jdbc:mysql://{{ .Values.db_host }}:{{ .Values.db_port }};databaseName={{ .Values.db_name }}" # 커넥터 접속 URL
        connection.password: "{{ .Values.db_pass }}"  # DB 유저 암호
        mode: incrementing  # 소스 커넥터의 동작 모드 
        incrementing.column.name: id  # incrementing 동작 기준 컬럼 
      # 커넥터별 설정 
      connectors:
      - name: "log1"   # 커넥터 이름. 여기서는 대상 DB 테이블 이름과 같게 하였음 
        config:
          topic.prefix: log1  # 카프카 토픽 접두사 
          query: "SELECT * FROM log1"  # 데이터를 가져올 쿼리문
      - name: "log2"
        config:
          topic.prefix: log2
          query: "SELECT * FROM log2"
    # DB 내 코드 테이블을 가져오기 위한 커넥터 그룹 
    - name: mydb_code  # 커넥터 그룹 이름 
      # 커넥터별 설정 
      connectors:
      # C_Code
      - name: "code"
        config:
          connector.class: io.confluent.connect.jdbc.JdbcSourceConnector    # 커넥터 클래스 
          tasks.max: "1"   # 커넥터 동시성 수 
          connection.url: "jdbc:mysql://{{ .Values.db_host }}:{{ .Values.db_port }};databaseName={{ .Values.db_name }}" # 커넥터 접속 URL
          connection.password: "{{ .Values.db_pass }}"  # DB 유저 암호
          mode: "bulk"        # 소스 커넥터의 동작 모드 
          topic.prefix: code  # 카프카 토픽 접두사 
          query: |            # 데이터를 가져올 쿼리문
            SELECT 
              CURRENT_TIMESTAMP() AS regtime,
              code, desc
            FROM code
```

갑자기 너무 많은 설정이 나와 당황스러울 것이나, 앞서 설명한 구조를 바탕으로 천천히 살펴보면 이해가 될 것이다.

`common` 및 `config` 에 기술된 각 커넥터 설정은 일종의 텍스트 템플릿이다. 텍스트내에 `{{ .Values.~ }}` 형식으로 기술하면 관련 변수가 앞선 `values` 에 있는 경우 대체된다. `{{ }}` 을 포함하지 않는 일반 텍스트는 그대로 커넥터 설정으로 사용된다.

커넥트 레벨, 커넥트 그룹 레벨의 `values` 블럭에서 기술된 변수는 결합되어 사용되며, 이때 중복된 변수 값이 있으면 가까운 값을 이용하게 된다. 예를 들어 다음과 같이 변수의 값이 지정되어 있다면 :

```yaml
connects:
  values:
    A: 1
  connector_groups:
  - name: group1
    values:
      A: 2
      B: 1
    connectors:
    - name: connector1
      config:
        A_is: {{ .Values.A }}
        B_is: {{ .Values.B }}
```

`connector1` 의 최종 설정에서 `A_is: 2`, `B_is: 1` 과 같게 된다. 

커넥터의 설정 값도 커넥터 그룹의 `common` 블럭 및 개별 커넥터의 `config` 블럭의 값이 같은 방식으로 결합되어 사용된다.

최종 결과물은 각 커넥터 별로 등록 가능한 쉘 스크립트 형식으로 저장되는데, 그것은 `[커넥트 타입]-[커넥터 그룹]-[커넥터 이름].sh` 형식 이름의 실행가능한 파일로 뒤에서 설명할 **초기화 파드** 에 저장된다. 

위 설정 예제의 경우 초기화 파드 `/usr/local/bin` 디렉토리 아래 다음과 같은 세가지 파일이 생성된다.

```
srccon-mydb_log-log1.sh
srccon-mydb_log-log2.sh
srccon-mydb_code-code.sh
```

각 파일의 내용은 다음과 같다. 

`srccon-mydb_log-log1.sh`
```
curl -s -X POST http://my-alpaka-srccon:8083/connectors -H "Content-Type: application/json" -d '{
    "name": "srccon-mydb_log-log1.sh",
    "config": {
        "connection.password": "mypass",
        "connection.url": "jdbc:mysql://mysql-db-addr;databaseName=mydb",
        "connector.class": "io.confluent.connect.jdbc.JdbcSourceConnector",
        "incrementing.column.name": "id",
        "mode": "incrementing",
        "query": "SELECT * FROM log1\n",
        "tasks.max": "1",
        "topic.prefix": "log1"
    }
}' | jq
```

`srccon-mydb_log-log2.sh`
```
curl -s -X POST http://my-alpaka-srccon:8083/connectors -H "Content-Type: application/json" -d '{
    "name": "srccon-mydb_log-log2.sh",
    "config": {
        "connection.password": "mypass",
        "connection.url": "jdbc:mysql://mysql-db-addr;databaseName=mydb",
        "connector.class": "io.confluent.connect.jdbc.JdbcSourceConnector",
        "incrementing.column.name": "id",
        "mode": "incrementing",
        "query": "SELECT * FROM log2\n",
        "tasks.max": "1",
        "topic.prefix": "log2"
    }
}' | jq
```

`srccon-mydb_code-code.sh`
```
curl -s -X POST http://my-alpaka-srccon:8083/connectors -H "Content-Type: application/json" -d '{
    "name": "srccon-mydb_code-code.sh",
    "config": {
        "connection.password": "mypass",
        "connection.url": "jdbc:mysql://mysql-db-addr;databaseName=mydb",
        "connector.class": "io.confluent.connect.jdbc.JdbcSourceConnector",
        "mode": "bulk",
        "query": "SELECT \n  CURRENT_TIMESTAMP() AS regtime,\n  code, desc\nFROM code\n\n",
        "tasks.max": "1",
        "topic.prefix": "code"
    }
}' | jq
```

잘 살펴보면 앞서 기술한 설정 및 변수값이 적용되어 있는 것을 알 수 있을 것이다. 이 파일이 실행되면 `curl` 을 통해 카프카 커넥트의 API 를 호출하여 각 커넥터를 등록하하게 된다.

> `test.enabled` 가 `true` 인 때에는 커넥터 등록 스크립트가 생성되지 않음에 주의하자. 

### 커넥터 등록과 갱신 

위에서 생성된 커넥터 등록 스크립트는 설치 완료 후 명시적으로 호출이 되어야 한다. 이를 위해서는 아래에서 설명할 **설치 후 초기화 설정** 을 이용하여야 하는데, 여기서 간단히 보여주면 아래와 같은 식이다.

```yaml
init:
  commands:
  - srccon-mydb_log-log1.sh
  - srccon-mydb_log-log2.sh
  - srccon-mydb_code-code.sh
```

그런데, 매번 이렇게 등록 스크립트를 기술해 주는 것은 불편하고 틀리기도 쉽다. 이에 모든 커넥터 등록 스크립트를 실행해주는 스크립트가 자동으로 생성되니 아래와 같이 이것을 이용하면 편리하다. 

```yaml
init:
  commands:
  - test-alpaka-srccon-all.sh
```

> `test.enabled` 가 `true` 인 경우 이 모든 커넥터 등록 스크립트가 생성되지 않으니 호출하지 않도록 하자.

이 스크립트의 파일명은 `[배포 이름]-alpaka-[커넥트 타입]-all.sh` 형식이다. 위 경우 `test` 배포에서 `srccon` 커넥트 아래에 만들어진 모든 등록 쉘스크립트가 이 안에 리스팅되어 실행되게 된다.

한 번 등록된 커넥터는 운영을 하면서 필요에 따라 설정을 바꿔야하는 경우도 빈번한데, 이 경우 지금까지 처럼 설정 파일에서 커넥터 설정을 바꿔주고 아래와 같이 Helm 의 업그레이드를 이용하면 적용된다.

```
helm upgrade -f configs/mymkb.yaml test alpaka/
```

일반적으로는 이렇게 하면 **설정파일 변경 -> ConfigMap 재생성 -> 관련 쿠버네티스 리소스 재생성** 식으로 진행되는데, 수정하지 않은 커넥트까지 불필요한 재시작을 겪게된다. 이에 알파카는 업그레이드시 사용된 커넥터 설정을 기 등록된 커넥터 설정과 비교하여 변경된 커넥터만 삭제 후 다시 등록하도록 구현되어 있다. 

### ksqlDB 설정 

[ksqlDB](https://ksqldb.io/) 는 SQL 형식 명령을 통해 카프카의 정보를 스트리밍 방식으로 처리하거나 질의할 수 있게 해준다. 일반적으로 다음과 같은 구성이다.

```yaml
ksqldb:
  enabled: false
  ksqldb:
    nameOverride: RELEASE-ksqldb
    kafka:
      # ksqlDB 차트에서 제공하는 카프카 사용 여부. 
      enabled: false 
      bootstrapServer: PLAINTEXT://RELEASE-kafka-headless:9092
    schema-registry: 
      # ksqlDB 차트에서 제공하는 스키마 레지스트리 사용 여부. 
      enabled: false 
    kafka-connect:
      # ksqlDB 차트에서 제공하는 커넥터 사용 여부. 
      enabled: false 
```

기본적으로 꺼져있는데, 지금가지 예제를 기준으로 사용하도록 설정한다면 다음과 같이 될 것이다.

```yaml
ksqldb:
  enabled: true 
  ksqldb:
    nameOverride: my-ksqldb
    kafka:
      enabled: false 
      bootstrapServer: PLAINTEXT://my-kafka-headless:9092  
```

ksqlDB 차트는 [여기](https://ricardo-aires.github.io/helm-charts/charts/ksqldb/) 에서 확인할 수 있다. 로컬 Helm 저장소에 등록하였다면 차트의 기본값은 다음처럼 확인할 수 있다.

```
helm show values rhcharts/ksqldb
```

### UI for Kafka 설정

[UI for Kafka](https://github.com/provectus/kafka-ui) 는 카프카 클러스터의 관리 및 모니터링에 사용된다. 이를 위해 연결할 카프카 브로커, 주키퍼 그리고 필요한 경우 커넥트 정보를 기술해주어야 한다. 일반적으로 다음과 같은 구성이다.

```yaml
ui4kafka:
  enabled: true           # UI for Kafka 를 이용하는 경우 true
  yamlApplicationConfig:  
    kafka:                # 연결할 카프카 클러스터 정보 
      clusters:
      - name: [배포 이름]-kafka   # 클러스터 이름 
        bootstrapServers: [배포 이름]-kafka-headless:9092  # 카프카 브로커의 헤드리스 서비스 
        zookeeper: [배포 이름]-zookeeper-headless:2181     # 주키퍼의 헤드리스 서비스 
        kafkaConnect:
        #
        # 이 아래 블록은 kafka_connect.connects 에 등록한 connect 들에 대해 기술
        # 커넥트 타입은 kafka_connect.connects.type 에 해당하는 값 
        #
        - name: [커넥트 타입]
          address: http://[배포 이름]-alpaka-[커넥트 타입]:8083
```

위의 커넥트 설정 예를 대상으로 한다면 다음과 같은 내용이 될 것이다.

```yaml
ui4kafka:
  enabled: true         
  yamlApplicationConfig:  
    kafka:            
      clusters:
      - name: my-kafka   # 카프카 클러스터 이름 
        bootstrapServers: my-kafka-headless:9092  # 카프카 브로커의 헤드리스 서비스 
        zookeeper: my-zookeeper-headless:2181     # 주키퍼의 헤드리스 서비스 
        kafkaConnect:
        # srccon 커넥트 정보 
        - name: srccon
          address: http://my-alpaka-srccon:8083
```

> `test.enabled` 가 `true` 인 경우 srccon 커넥트의 주소는 `address: http://my-alpaka-test-srccon:8083` 를 이용해야 한다.

UI for Kafka 차트는 최초에 [Provectus 의 것](https://github.com/provectus/kafka-ui)을 이용하였으나, Helm 패키지 설치에 문제가 있어 [포크한 것](https://github.com/haje01/kafka-ui) 을 이용하였다. 로컬 Helm 저장소에 등록하였다면 차트의 기본값은 다음처럼 확인할 수 있다.

```
helm show values kafka-ui/kafka-ui
```

### 쿠버네티스 대쉬보드 설정 

[쿠버네티스 대쉬보드](https://github.com/kubernetes/dashboard)는 쿠버네티스 클러스터 자체의 모니터링을 위해 사용한다. 설정은 대략적으로 다음과 같은 구조를 가진다. 

```yaml
k8dashboard:
  enabled: true                 # 쿠버네티스 대쉬보드를 이용하는 경우 true
  protocolHttp: true 
  # 서비스 관련 
  service:
    externalPort: 8443
  serviceAccount:
    name: k8dash-admin
  # 간편한 로그인
  extraArgs:                    
    - --token-ttl=86400
    - --enable-skip-login       
    - --enable-insecure-login
```

대개의 경우 기본 값을 이용하면 되기에 다음 처럼 간단히 `enabled` 만 설정한다.

```yaml
k8dashboard:
  enabled: true 
```

쿠버네티스 대쉬보드의 차트는 [여기](https://artifacthub.io/packages/helm/k8s-dashboard/kubernetes-dashboard) 에서 찾을 수 있다. 로컬 Helm 저장소에 등록하였다면 차트의 기본값은 다음처럼 확인할 수 있다.

```
helm show values kubernetes-dashboard/kubernetes-dashboard 
```

### 프로메테우스 설정

[프로메테우스](https://prometheus.io/) 는 서버 및 서비스 관련 메트릭을 모니터링하는데 사용한다. 설정은 대략적으로 다음과 같은 구조를 가진다. 

```yaml
prometheus:
  enabled: true                  # 프로메테우스를 이용하는 경우 true
  prometheus:
    enabled: true
    # 메트릭 수집 관련 
    additionalScrapeConfigs:
      enabled: true
      type: internal
      internal:
        # 수집 잡 리스트
        jobList:
        - job_name: [잡 이름]

CHANGES.md
> 여기에 버전업 후 변화 내용을 기록하고 tag 도 함께 생성한다.

## 0.0.5 
- deploy 설정을 submodule 로 추가 
- multi connector type 버그 수정

## 0.0.4
- Loki 적용

## 0.0.3
- 변화 기록 시작



tmux-portfwd.sh
session="portfwd"

tmux new-session -d -s $session

# K8S Dashboard
window=0
tmux rename-window -t $session:$window 'k8s-dashboard'
tmux send-keys -t $session:$window 'export K8DASH_POD=$(kubectl get pods -l "app.kubernetes.io/component=kubernetes-dashboard" -A -o jsonpath="{.items[0].metadata.name}")' C-m
tmux send-keys -t $session:$window 'kubectl port-forward $K8DASH_POD -n default 8443:8443' C-m

# UI for Kafka
window=1
tmux new-window -t $session:$window -n 'kafka-ui'
tmux send-keys -t $session:$window 'export KUI_SVC=$(kubectl get svc -l "app.kubernetes.io/name=kafka-ui" -A -o jsonpath="{.items[0].metadata.name}")' C-m
tmux send-keys -t $session:$window 'kubectl port-forward svc/$KUI_SVC 8989:80' C-m

# Prometheus
window=2
tmux new-window -t $session:$window -n 'prometheus'
tmux send-keys -t $session:$window 'export PROM_SVC=$(kubectl get svc -l "app.kubernetes.io/component=prometheus" -A -o jsonpath="{.items[0].metadata.name}")' C-m
tmux send-keys -t $session:$window 'kubectl port-forward svc/$PROM_SVC 9090:9090' C-m

# Grafana
window=3
tmux new-window -t $session:$window -n 'grafana'
tmux send-keys -t $session:$window 'export GRAFA_SVC=$(kubectl get svc -l "app.kubernetes.io/name=grafana" -A -o jsonpath="{.items[0].metadata.name}")' C-m
tmux send-keys -t $session:$window 'kubectl port-forward svc/$GRAFA_SVC 3000' C-m

tmux attach-session -t $session
